### ISSUE TITLE
Convert EIA-861  ETL to use Dagster

### ISSUE THREAD

**Issue by bendnorman** [catalyst-cooperative/pudl#2265]
Create dagster assets for the raw and transformed EIA 861 data so the partially cleaned tables can be accessed in the pudl.sqlite DB. 

The EIA 861 ETL currently lives in a method in the PudlTabl class instead of the ETL portion of PUDL because the harvesting needs to be refactored to accommodate harvesting new data sources. See #639 and #614.

Once the ETL has been converted to dagster #2104 we can start to persist partially cleaned tables from new data sources.

```[tasklist]
- [x] Add `raw_eia861` @multi_asset in `pudl.extract.eia861`.
- [x] Identify which EIA-861 transform functions are `@asset` vs. `@multi_asset` and mark them.
- [x] Update EIA-861 transform functions to be `@asset` or `@multi_asset`.
- [x] Remove `pudl.transform.eia861.transform()`
- [x] Apply `_early_transform()` to all tables before per-table transformations.
- [x] Enforce PUDL data types after all transforms are complete.
- [x] Convert `balancing_authority_assn_eia861` to take all data tables as input.
- [x] Convert `utility_assn_eia861` to take all data tables as input.
- [x] Verify that we can materialize all simple assets using default IO Manager
- [x] Define table schema for `utility_assn_eia861`
- [x] Define table schema for `balancing_authority_assn_eia861`
- [x] Define table schema for `balancing_authority_eia861`
- [x] Define table schema for `service_territory_eia861`
- [x] Define table schema for `sales_eia861`
- [x] Fix transform problems arising from introduction of `early_release` column.
- [x] Load `utility_assn_eia861` into SQLite
- [x] Load `sales_eia861` into SQLite
- [x] Load `balancing_authority_eia861` into SQLite
- [x] Load `balancing_authority_assn_eia861` into SQLite (replaced NA `state` w/ `UNK`)
- [x] Load `service_territory_eia861` into SQLite (Fixed `county_id_fips` & some duplicate PKs)
- [x] Remove the `PudlTabl.etl_eia861` method.
- [x] Disable PKs and attempt to load other clean tables into the DB
- [x] Update `test_eia861_etl` to check that EIA-861 outputs work and are non-null.
- [x] Add `eia861` sanity checks. Currently: No all null columns, PKs are unique & not null.
- [x] Update `*_eia861` methods in PudTabl to read from PUDL DB.
- [x] Disable automatic FK constraints in EIA-861 tables as needed (these aren't finished tables)
- [x] Make sure that the `pudl.output.ferc714` routines that depend on EIA-861 data still work.
- [x] Make sure that the state demand allocation analysis still works.
- [x] ~Make sure that `service_territories` still works.~ Oh right, it's broken on `dev` too: Memory 💣 
- [x] Fix missing most-recent-year issue with discontinued EIA-861 tables in CI (process all years!)
- [x] Add Release Notes covering new EIA-861 data.
- [x] Add an EIA-861 data source template to the docs build.
```

## Out of Scope

- Adding the EIA-861 tables to the harvesting process or `pudl.eia.eia_transform()`.
- Integrating additional EIA-861 tables besides the 5 above to the database.
- Persisting any of the raw EIA-861 tables in the DB. 
- Fixing PK NOT NULL / UNIQUE or FK constraint errors in any tables other than `sales_eia861`, `service_territory_eia861`, `balancing_authority_eia861`, `balancing_authority_assn_eia861`, `utility_assn_eia861`.

## New Issues

Some stuff that came up while addressing this issue, but that was out of scope:

* #2412
* #2414 
* #2417 
* #2418 
* #2419 
* #2420

**Comment by zaneselvans** [comment]
@bendnorman How should we provide access to the tables that aren't going to get written to the SQLite DB? Only a few of them (listed above) have been used in other analyses by us, but in theory there could be folks out there who have been relying on being able to access them through `PudlTabl`.
* Compiling Resource metadata for all of them would be a fair bit of work. We have to do it eventually so maybe we just... do it?
* Is there some dumb/fast way to write these tables into the DB?
* Do we go to only having these tables available from within Dagster?

**Comment by zaneselvans** [comment]
@bendnorman A couple more questions:
* There's a little pre-processing function (`_early_transform()`) that should be applied to all of the tables coming into the EIA-861 transforms. Before it was applied to each of them in turn while looping over them all, but that doesn't happen now. Is there an easy way to replace that? Or does that function need to be called in every transform function now?
* Similarly we need to enforce data types on all of them after the transforms are done.
* There are a couple of transform functions (that are more like harvesting functions) that depend on almost every transformed table, but output just a single table each. Is there an easy way to specify that kind of dependency, other than manually enumerating the 20 different asset IDs in the call signature? Is the `eia_transform()` asset the best example of this kind of thing to work off of?

**Comment by zaneselvans** [comment]
I used asset factories to apply uniform pre- and post-processing to the EIA-861 assets. It works, and shows the real dependencies, but there's a lot of startup overhead for each of the little pre/post cleaning operations so I imagine just running everything through a multi-asset before and a multi-asset after is probably faster.

But hey it all works!

![image](https://user-images.githubusercontent.com/596279/225164080-5f920cb7-2da7-4b43-aea2-3935f64d6fd9.png)

Now I need to define metadata for the 5 tables that are done enough to go into the DB and update their IO Managers.

**Comment by zaneselvans** [comment]
Okay rather than using the asset factories, I just integrated the pre- and post-processing into the individual table transformers and it's much simpler & faster, and still keeps the dependencies very clear:

![image](https://user-images.githubusercontent.com/596279/225229574-82531ff0-8673-42bd-b846-4cb77a2ce9be.png)

**Comment by zaneselvans** [comment]
When I attempt to store a table using the SQLite IO Manager, but the dataframe has (say) some extra columns the Dagster run still appears to succeed, which seems wrong.

Then if I attempt to read the table out of the database later I get errors like:
```
ValueError: Couldn't parse datetime string: '2001-01-01'
ValueError: balancing_authority_eia861 not found. Either the table was dropped or it doesn't exist in the pudl.metadata.resources. Add the table to the metadata and recreate the database.
```

Need to break this down one table at a time in more detail.

**Comment by cmgosnell** [comment]
❌ balancing_authority_assn_eia861

* do these records have unique non-state primary keys?
* do the ba's in question have states that show up in other tables? ie could you infer the state from the rest of the data? (this would introduce ETL coupling so would maybe be a better output layer step) 
* I'm pro `UNK` over drop or breaking them out into a separate table

❌ service_territory_eia861

* why is the FIPS ID a part of the PK? I assume it would be the nicest if it could be, but it sounds like it should really be the native state/county/zip data.
* If keeping the FIPS ID as the pk is a priority, are there FIPS-equivalent keys for non-US geographies we could add?
* Don't we have county name cleaning somewhere lying around? Cleaning in a totally standard way may not be possible but keeping a list of known misspellings mapped to their true county name seems like a resource we could use in a bunch of places and add to.

**Comment by zaneselvans** [comment]
Thanks for that great suggestion on `balancing_authority_assn_eia861`! Unfortunately it turns out that all of the records with NA states have unique combinations of date, utility, and balancing authority, so we can't just drop them without losing information. So for now I think I'll just `UNK` them.

After a little messing around with the Census data, the county names are very fixable. There's only 1 "county" in Guam, and 4 in the Virgin Islands, and a couple dozen in AK, and the rest are misspellings I can easily add to our existing dictionary of fixes. So I'll go ahead with that.

**Comment by zaneselvans** [comment]
I might be doing something wrong in my exploration, but it also seems like the records that have NA values for states are combinations of BA + Util IDs that **never** have any states associated with them (I briefly looked at forward/back filling) so now I feel like maybe they are truly spurious and maybe should be dropped. But we can save that for another time.

**Comment by zaneselvans** [comment]
I've (temporarily) added all years of `eia861` data into the fast ETL settings, since discontinued tables currently cause issues when running just the 2020 or 2021 data. The EIA-861 is small and runs in about a minute even with all data (and it was all getting processed in the tests before...) so this isn't so bad, and will keep the CI useful in the meantime.

**Comment by zaneselvans** [comment]
## Some Dagster Dev Notes

I found myself frequently needing to delete `pudl.sqlite` and often only realizing it after some unexpected behavior and time spent debugging and being confused. E.g. when relaxing the FK constraints I re-ran the ETL and checked again whether they were being violated, but nothing had changed! (b/c re-running the ETL doesn't update the DB schema). 

Needing to switch an asset back to being written to disk after discovering a problem with loading it into the DB in order to inspect it and see what was going on felt a little annoying.

Needing to run the FK checks outside of the ETL from the command line seemed incongruous. There must be some way to integrate that into the ETL?

I found myself unsure of when I needed to reload what and so became paranoid about resetting everything frequently rather than waste time debugging something that was just a stale environment. It would be good to document when each of these actions are necessary, so that only the least disruptive thing has to get done.
* Reloading the code location.
* Reloading the "defs"
* Deleting `pudl.sqlite` from disk
* Restarting dagster / dagit
* Reloading the kernel in my development notebook.
* Materializing an individual asset, or all failed assets, or all stale & missing assets.

The times when I could make a change, and then just re-materialize the asset it applied to were amazing!  Or when one thing would fail, and I could change the code, and re-run just that asset and everything downstream from it. Super convenient! But uncertainty about when that was the right thing to do, and when I needed to start everything over from scratch meant I probably didn't take advantage of that potential efficiency as often as I could have.
